Remove toms748 bound-tweaking

fmahebert opened the pull request:
## Proposed changes

1. Remove the bounds tweaks in `toms748`, as per discussion in #439. Make corresponding changes to `toms748` tests.
1. Add note of how bounds are handled in `newton_raphson`.

### Types of changes:

- [x] Bugfix

### Component:

- [x] Code

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

fmahebert committed:
Remove bounds-tweaking in TOMS748 root finder

fmahebert commented:
Clang-tidy is failing because of boost stuff, but the other two failures are on me...

wthrowe commented:
I have checked the clang-tidy failures and they should be ignored.

kidder reviewed (commented):
squash

wthrowe reviewed (dismissed):
clang-tidy failures are still irrelevant

nilsdeppe commented on src/NumericalAlgorithms/RootFinding/NewtonRaphson.hpp:
```diff
@@ -22,6 +22,10 @@ namespace RootFinder {
  *
  * \snippet Test_NewtonRaphson.cpp double_newton_raphson_root_find
```
Just drop this commit. People can read the boost documentation.

wthrowe replied:
I couldn't find anything stating this in the Boost docs, but that's probably because it doesn't seem to be true.  If I modify the test to put the lower bound `1e-10` above the correct result (requesting 8 digits), I do not get the correct answer to 8 digits.

So yes, please drop this.

fmahebert replied:
I had modified the test and determined that lowering the upper bound did not break the algorithm, but then I incorrectly generalized. I agree that raising the lower bound _does_ break the algorithm. ¯\\\_(ツ)\_/¯

I will drop this.

nilsdeppe commented on src/NumericalAlgorithms/RootFinding/TOMS748.hpp:
```diff
@@ -68,6 +68,11 @@ double toms748(const Function& f, const double lower_bound,
  *
  * \snippet Test_TOMS748.cpp datavector_root_find
  *
+ * The upper and lower bounds must bracket the root so that
```
Why not just say that the root find searches in (a, b) not [a, b]?

fmahebert replied:
I went looking in boost... their documentation says

> It is a precondition that ... `f(a) * f(b) < 0`

but their source code reads `if ((fa == 0) || (fb == 0)) { return one of a or b }`, which would be better matched by `<=` in the documentation.

I suggest saying, based on your comment but making it consistent with the boost code,

"The TOMS748 algorithm searches for a root in the interval [`lower_bound`, `upper_bound`], and will throw an exception if this interval does not bracket a root, i.e. if `f(lower_bound) * f(upper_bound) > 0`."

nilsdeppe commented on tests/Unit/NumericalAlgorithms/RootFinding/Test_TOMS748.cpp:
```diff
@@ -44,31 +44,69 @@ SPECTRE_TEST_CASE("Unit.Numerical.RootFinding.TOMS748.Bounds",
+    INFO("Expected rootfinder to fail because lower bound is too tight");
+    CHECK(false);
+  } catch (...) {
+    exception_count++;
```
Don't do this, check that the exception thrown is what you expect.

nilsdeppe commented on tests/Unit/NumericalAlgorithms/RootFinding/Test_TOMS748.cpp:
```diff
@@ -44,31 +44,69 @@ SPECTRE_TEST_CASE("Unit.Numerical.RootFinding.TOMS748.Bounds",
-  double upper = 2.0;
-  double lower = sqrt(2.0);
+  const double upper = 2.0;
+  const double lower = 0.0;
```
why 0.0? Don't you just want `sqrt(2.0)-epsilon`?

nilsdeppe commented on tests/Unit/NumericalAlgorithms/RootFinding/Test_TOMS748.cpp:
```diff
@@ -44,31 +44,69 @@ SPECTRE_TEST_CASE("Unit.Numerical.RootFinding.TOMS748.Bounds",
+  CHECK(std::abs(root - sqrt(2.0)) / sqrt(2.0) < rel_tol);
+
+  // Check that close but correct bounds work
+  {
```
Don't scope everything, just inline the variables.

nilsdeppe commented on tests/Unit/NumericalAlgorithms/RootFinding/Test_TOMS748.cpp:
```diff
@@ -44,31 +44,69 @@ SPECTRE_TEST_CASE("Unit.Numerical.RootFinding.TOMS748.Bounds",
   const double rel_tol = 1e-15;
-  const DataVector upper{2.0, 3.0, -sqrt(2.0), -sqrt(2.0)};
-  const DataVector lower{sqrt(2.), sqrt(2.0), -2.0, -3.0};
+  const DataVector upper{2.0, 3.0, -1.0, 1.0};
```
Again, set a more appropriate bound for what you're testing.

fmahebert commented:
I've removed the commit re: Newton Raphson behavior.

~I've also added ASSERTs against a possible failure mode (lack of convergence) of the root finders... but if we don't want this change as part of this PR I can make a different PR.~

Edit: commit removed after discussion below, will become a new PR.

nilsdeppe commented on src/NumericalAlgorithms/RootFinding/NewtonRaphson.hpp:
```diff
@@ -41,9 +41,11 @@ double newton_raphson(const Function& f, const double initial_guess,
+  auto result = boost::math::tools::newton_raphson_iterate(  // NOLINT
       f, initial_guess, lower_bound, upper_bound,
       std::round(std::log2(std::pow(10, digits))), max_iters);
+  ASSERT(max_iters < max_iterations, "Root finder did not converge.");
```
You should throw instead since we already allow exceptions from boost to propagate and be caught. This will also allow us to have a backup plan in case something doesn't converge.

fmahebert replied:
Sounds good. Do you have an opinion on the type of exception to throw? Perhaps `std::runtime_error`? Or a custom "convergence failure" exception?

nilsdeppe replied:
I don't, though I do see a benefit to having a custom exception that could be checked for. @wthrowe @kidder?

kidder replied:
A custom exception sounds good to me for algorithmic failures that we potentially can work around with a backup plan

fmahebert replied:
Okay. I like where this idea is heading, but it's also growing into a larger change that IMO should be a different PR from this bugfix. I'll remove the commit from here and open a new PR when ready.

nilsdeppe replied:
I'd be fine with it being a separate commit or a separate PR, whatever you prefer :)

fmahebert replied:
I'll make a separate PR. I've removed the relevant commit from this PR, so it should be ready for you to look at my fixups to your previous review :)

nilsdeppe commented on tests/Unit/NumericalAlgorithms/RootFinding/Test_TOMS748.cpp:
```diff
@@ -54,59 +55,40 @@ SPECTRE_TEST_CASE("Unit.Numerical.RootFinding.TOMS748.Bounds",
-  } catch (...) {
-    exception_count++;
+  } catch (std::domain_error&) {
   }
```
And then:
```
} catch (...) {
  CHECK(false);
}
```
to make sure we don't propagate an incorrect exception. If you know what the exception message should be it would be good to check that too :)

Apply the same for your other checks of this type

nilsdeppe reviewed (changes requested):
I'm happy for you to squash as you do this :)

kidder commented:
@nilsdeppe please look at

kidder commented:
travis failure is from boost
